contracts/test/TestUniswap.sol

```diff
@@ -0,0 +1,71 @@
+pragma solidity ^0.5.16;
+
+import "openzeppelin-solidity/contracts/token/ERC20/IERC20.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+import "../paymaster/IUniswap.sol";
+
+contract TestToken is ERC20 {
+    function mint(uint amount) public {
+        _mint(msg.sender, amount);
+    }
+}
+
+//naive, no-calculation swapper.
+//- the exchange rate is fixed at construction
+//- mints new tokens at will...
+contract TestUniswap is IUniswap {
+    IERC20 public token;
+    uint public rateMult;
+    uint public rateDiv;
+
+    constructor(uint _rateMult, uint _rateDiv) public payable {
+        token = new TestToken();
+        rateMult = _rateMult;
+        rateDiv = _rateDiv;
+        require(msg.value > 0, "must specify liquidity");
+        require(rateMult != 0 && rateDiv != 0, "bad mult,div");
+    }
+
+    function() external payable {}
+
+    function tokenAddress() external view returns (address out) {
+        return address(token);
+    }
+
+    function addLiquidity(uint256 min_liquidity, uint256 max_tokens, uint256 deadline) external payable returns (uint256 out) {
+        min_liquidity;
+        max_tokens;
+        deadline;
+        return 0;
+    }
+
+    event SwapTokensToEth(address msgsender, uint tokensToSell, uint tokAllowance, uint eth_bought, uint ethbal, int tokbal);
```

Comment:
Would it be helpful to define the events in the interface file?

Reply:
obsolete event anyway...